lib/spack/llnl/util/tty/log.py

```diff
@@ -33,8 +33,21 @@
     pass
 
 
+esc, bell, lbracket, bslash, newline = r"\x1b", r"\x07", r"\[", r"\\", r"\n"
+# Ansi Control Sequence Introducers (CSI) are a well-defined format
+csi_pre = f"{esc}{lbracket}"
+csi_param, csi_inter, csi_post = r"[0-?]", r"[ -/]", r"[@-~]"
+ansi_csi = f"{csi_pre}{csi_param}*{csi_inter}*{csi_post}"
+# General ansi escape sequences have well-defined prefixes,
+#  but content and suffixes are less reliable.
+# Conservatively assume they end with either "<ESC>\" or "<BELL>",
+#  with no intervening "<ESC>"/"<BELL>" keys or newlines
```

Comment:
Is there a good reference here?

I am not a terminal control character expert :) but I imagine the worst case scenario here is that an escape sequence isn't terminated (which appears to be allowed) and we throw away the rest of the line. Is that a correct interpretation?

Reply:
> Is there a good reference here?

I'm also not an expert, so I can't point you to a section of the official standard doc (I assume whoever chose to represent characters by just the codes in the ECMA-48 standard you linked is a mad genius). 

But looking at the first table under "FE Escape sequences" here, https://en.m.wikipedia.org/wiki/ANSI_escape_code#Fe_Escape_sequences you can see the "String Terminator" code is `<ESC>\` and most of the listed examples in the rest of that table are terminated by the string terminator.

And some sequences can instead be terminated by `<BELL>` on xterm https://en.m.wikipedia.org/wiki/ANSI_escape_code#OSC_(Operating_System_Command)_sequences

> I imagine the worst case scenario here is that an escape sequence isn't terminated (which appears to be allowed) and we throw away the rest of the line. Is that a correct interpretation?

The worst case is just that we leave the unterminated sequence in the log file, since the regex match requires finding the introducer and the terminator without any unexpected characters like newlines or other escapes in between. So this won't catch all possible codes, but catches more and is safer in the case of unexpected formats